pkg/querier/querier.go

```diff
@@ -1144,6 +1145,62 @@ func (q *SingleTenantQuerier) DetectedFields(ctx context.Context, req *logproto.
 	}, nil
 }
 
+func (q *SingleTenantQuerier) StructuredMetadata(ctx context.Context, req *logproto.StructuredMetadataKeysRequest) (*logproto.StructuredMetadataKeysResponse, error) {
+	expr, err := syntax.ParseLogSelector(req.Query, true)
+	if err != nil {
+		return nil, err
+	}
+	params := logql.SelectLogParams{
+		QueryRequest: &logproto.QueryRequest{
+			Start:     req.Start,
+			End:       req.End,
+			Limit:     1000,
+			Direction: logproto.BACKWARD,
+			Selector:  expr.String(),
+			Plan: &plan.QueryPlan{
+				AST: expr,
+			},
+		},
+	}
+
+	iters, err := q.SelectLogs(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(twhitney): we can improve this in the future as we don't need to iterate lines for structured metadata, just chunks
+	streams, err := streamsForFieldDetection(iters, 1000)
+	if err != nil {
+		return nil, err
+	}
+
+	structuredMetadata := parseStructuredMetadata(ctx, streams)
+
+	return &logproto.StructuredMetadataKeysResponse{
+		Keys: structuredMetadata,
+	}, nil
+}
+
+func parseStructuredMetadata(ctx context.Context, streams logqlmodel.Streams) []string {
+	lbls := []string{}
+
+	for _, stream := range streams {
+		level.Debug(spanlogger.FromContext(ctx)).Log(
```

Comment:
I think you need to remove that line. Too intense and it seems to also span log it.

Reply:
Also makes sense. Should I remove https://github.com/grafana/loki/blob/main/pkg/querier/querier.go#L1217-L1219 as well?

Reply:
Yes please